File: DSA/Hima-Swarupa-Padala/day19q1.py
from collections import defaultdict, deque

class Graph:
    def __init__(self, nodes):
        self.graph = defaultdict(list)
        self.nodes = nodes

    def add_edge(self, u, v):
        self.graph[u].append(v)
        self.graph[v].append(u)

    def bfs(self, start):

        distances = {}
        for i in range(1, self.nodes + 1):
            distances[i] = -1
        visited = [False] * self.nodes

        queue = deque()
        queue.append(start)
        distances[start] = 0
        visited[start - 1] = True

        while queue:
            node = queue.popleft()
            for neighbor in self.graph[node]:
                if not visited[neighbor - 1]:
                    queue.append(neighbor)
                    visited[neighbor - 1] = True
                    distances[neighbor] = distances[node] + 1

        return distances

def main():

    nodes = 7
    edges = [(1, 2), (1, 3), (2, 4), (2, 5), (3, 6), (3, 7)]
    start_node = 1

    graph = Graph(nodes)
    for u, v in edges:
        graph.add_edge(u, v)

    shortest_reach = graph.bfs(start_node)

    print("Shortest reach from starting node:")
    print(shortest_reach)

if __name__ == "__main__":
    main()

File: DSA/Yashvardhan-Guliyani/Day19_q1.java
import java.util.*;

public class ShortestReachBFS {
    private int nodes;
    private List<List<Integer>> adjacencyList;

    public ShortestReachBFS(int n) {
        this.nodes = n;
        this.adjacencyList = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            adjacencyList.add(new ArrayList<>());
        }
    }

    public void addEdge(int u, int v) {
        adjacencyList.get(u).add(v);
        adjacencyList.get(v).add(u);
    }

    public int[] shortestReach(int start) {
        int[] distances = new int[nodes];
        Arrays.fill(distances, -1);
        distances[start] = 0;

        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);

        while (!queue.isEmpty()) {
            int current = queue.poll();

            for (int neighbor : adjacencyList.get(current)) {
                if (distances[neighbor] == -1) {
                    distances[neighbor] = distances[current] + 1;
                    queue.offer(neighbor);
                }
            }
        }

        return distances;
    }

    public static void main(String[] args) {
      
        int n = 6; 
        int m = 8; 
        int[][] edges = {{0, 1}, {0, 2}, {1, 2}, {1, 3}, {2, 3}, {3, 4}, {2, 4}, {4, 5}};
        int start = 0; 
        
        ShortestReachBFS graph = new ShortestReachBFS(n);

        for (int[] edge : edges) {
            graph.addEdge(edge[0], edge[1]);
        }

        int[] result = graph.shortestReach(start);

        System.out.println("Shortest reach from node " + start + " to all other nodes:");
        for (int i = 0; i < n; i++) {
            System.out.println("Node " + i + ": " + result[i]);
        }
    }
}
